app/lib/analyzer/scheduler.dart

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+/// Schedules and executes package analysis.
+class AnalysisScheduler {
+  final Stream taskReceivePort;
+  final List<ScheduledTask> _tasks = [];
+  final Set<ScheduledTask> _triggered = new Set();
+  StreamSubscription _subscription;
+  DateTime _lastDatastorePoll;
+
+  AnalysisScheduler(this.taskReceivePort);
+
+  Future run() async {
+    _subscription = taskReceivePort.listen((ScheduledTask task) {
+      _triggered.add(task);
+    });
+    await _runLoop();
+    // It is unlikely that we exit the run loop, but cancelling the subscription
+    // makes dartanalyzer happy.
+    await _subscription.cancel();
+  }
+
+  Future _runLoop() async {
+    for (;;) {
+      if (_tasks.isEmpty) {
+        await _populateTasks();
+        // short nap, just in case
+        await new Future.delayed(new Duration(seconds: 2));
+      } else {
+        final ScheduledTask task = _tasks.removeLast();
+        await _runAnalyze(task);
+        // long nap, to give GC ample time
+        await new Future.delayed(new Duration(seconds: 15));
```

Comment:
Actually I'm not sure if our GC is collecting during idle times. Usually the collections are triggered by new/old-space allocation failures (i.e. if current capacity is exhausted). But I don't mind if we are waiting for a few seconds :)